### ISSUE TITLE
Scan-level memory bottleneck on optimize()/forward(): need segmented remat for large n_steps

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#73]
## Summary

`Simulation.forward(checkpoint=True)` (and therefore `optimize()`) currently applies `jax.checkpoint` at the **per-step** level only:

```python
# rfx/simulation.py:1027
body = jax.checkpoint(step_fn) if checkpoint else step_fn
final_carry, outputs = jax.lax.scan(body, carry_init, xs)
```

`jax.checkpoint(step_fn)` rematerialises *within* a single step, but `jax.lax.scan` still keeps the **full carry at every step** for the backward pass. With ~30 MB carry on a ~0.4 M-cell uniform grid (state + CPML aux + DFT accumulators), peak forward memory grows linearly with `n_steps`:

| n_steps | scan carry memory | A6000 (48 GB)? |
|---:|---:|:--|
|  800 | 24 GB | OK |
| 1500 | 45 GB | borderline |
| 3000 | 90 GB | **OOM** (observed 79.7 GB peak; HLO remat unable to reduce) |

Concrete failure: rfx-tap paper, ex1_dra at 5 GHz with `optimize(n_steps=3000)` — required for the wave-decomp DFT (#72) to capture full ringdown of a Q≈30-50 DRA — OOMs immediately on a single A6000:

```
W external/xla/xla/service/hlo_rematerialization.cc:3005] Can't reduce memory use below
4.10 GiB by rematerialization; only reduced to 79.67 GiB ...
RESOURCE_EXHAUSTED: Out of memory while trying to allocate 11370592136 bytes.
```

## Knock-on physical effect (why this blocks #72's actual goal)

The wave-decomposition |S11| objective from #72 needs `n_steps` long enough that the V/I DFT integrates past the antenna's ringdown. For a Q=50 DRA at 5 GHz the ringdown is ~3 ns ≈ 1500 steps, ~5τ ≈ 8 ns ≈ 4000 steps. With memory capping `n_steps` at ~800, the DFT is dominated by the source-driven transient (V/I look matched to a 50 Ω source regardless of the actual load), so the optimiser converges to a "fake match" that post-eval (with longer integration) reveals as |S11| ~ -3.5 dB instead of < -10 dB. Solving #72's formula was necessary but not sufficient — the integration window has to be long enough as well.

## Proposed fix — segmented remat

Wrap an **inner scan over a segment** with `jax.checkpoint`, and run an **outer scan over segments**. With segment size `s = √n_steps`, peak forward memory drops from `O(n_steps · |carry|)` to `O(√n_steps · |carry|)`:

| n_steps | linear (current) | segmented (√) |
|---:|---:|---:|
|  800 |  24 GB | 1.6 GB |
| 1500 |  45 GB | 2.3 GB |
| 3000 |  90 GB | 3.3 GB |
| 6000 | 180 GB | 4.6 GB |

API additions (opt-in, default off → no behavioural change):

- `Simulation.forward(..., checkpoint_segments: int | None = None)`
- `optimize(..., checkpoint_segments: int | None = None)` — passes through

Sketch:

```python
# rfx/simulation.py
def run(..., checkpoint=False, checkpoint_segments=None, ...):
    if checkpoint_segments is None:
        body = jax.checkpoint(step_fn) if checkpoint else step_fn
        final_carry, outputs = jax.lax.scan(body, carry_init, xs)
    else:
        K = int(checkpoint_segments)
        s = (n_steps + K - 1) // K
        # Pad xs to K*s if needed
        xs_padded = _pad_xs(xs, K * s)
        xs_segmented = xs_padded.reshape(K, s, *xs.shape[1:])
        def segment_body(carry, seg_xs):
            new_carry, seg_outputs = jax.lax.scan(step_fn, carry, seg_xs)
            return new_carry, seg_outputs
        seg_body_ckpt = jax.checkpoint(segment_body, prevent_cse=False)
        final_carry, seg_outputs = jax.lax.scan(seg_body_ckpt, carry_init, xs_segmented)
        outputs = jax.tree.map(lambda x: x.reshape(K * s, *x.shape[2:])[:n_steps], seg_outputs)
```

DFT-style accumulators (lumped/wire/waveguide port S-params) live in the `carry`, so they thread through segment boundaries automatically.

`prevent_cse=False` is required so that the inner scan is not de-duplicated across the outer scan — without it, JAX may decide it can re-trace the segment cheaply and undo the memory savings.

## Scope

- Phase A — uniform path (`run`, `run_until_decay` in `simulation.py`).
- Phase B — `Simulation.forward` and `_forward_from_materials` plumb-through.
- Phase C — `optimize` plumb-through.
- Phase D — tests: `tests/test_scan_segmented_checkpoint.py`
  - segmented forward output bit-exact (or atol 1e-5) vs unsegmented for the same n_steps,
  - segmented `jax.grad` matches unsegmented to atol 1e-5,
  - DFT accumulator (lumped port s_params) matches across segments.

NU path is out of scope for this PR (different forward function); it can mirror the same pattern in a follow-up.

## Workload

~150–200 LoC core + ~150 LoC tests. Volunteering for the PR. Related: #72 (wave-decomposition path that this unblocks for high-Q antennas).

**Comment by bk-squared** [comment]
Fixed by **b6c4153** `feat(forward): segmented JAX scan checkpointing` (on origin/main). Adds opt-in `forward(checkpoint_segments=K)` / `optimize(checkpoint_segments=K)` knob splitting the n_steps scan into K segments of size √n_steps for ~√n memory at ~2× compute cost. Uniform single-device path only; NU and distributed=True raise NotImplementedError. Tests: tests/test_scan_segmented_checkpoint.py 6/6 pass (forward+gradient equivalence vs unsegmented, divisor enforcement, NotImplementedError guards).
